CHANGELOG.md

```diff
@@ -1,5 +1,108 @@
+- Support annotated `self` parameter in constructor of non-invariant generic classes ([#21325](https://github.com/astral-sh/ruff/pull/21325))
+- Improve generic constructor inference ([#20933](https://github.com/astral-sh/ruff/pull/20933))
+- Improve generic call expression inference ([#21210](https://github.com/astral-sh/ruff/pull/21210))
+- Improve generic class constructor inference ([#21442](https://github.com/astral-sh/ruff/pull/21442))
+- Improve literal promotion heuristics ([#21439](https://github.com/astral-sh/ruff/pull/21439))
+- Improve conditional expression inference ([#21443](https://github.com/astral-sh/ruff/pull/21443))
+- Make `__getattr__` available for `ModuleType` instances ([#21450](https://github.com/astral-sh/ruff/pull/21450))
+- Introduce implicit local variables for `from` imports of submodules in `__init__.py(i)` ([#21173](https://github.com/astral-sh/ruff/pull/21173))
+- Make implicit submodule locals only occur in global scope of an `__init__.py(i)` ([#21370](https://github.com/astral-sh/ruff/pull/21370))
+- Make implicit submodule locals also occur for absolute `from` imports in `__init__.py(i)` files ([#21372](https://github.com/astral-sh/ruff/pull/21372))
+- Consider `from thispackage import y` a re-export of `y` in `__init__.pyi` ([#21387](https://github.com/astral-sh/ruff/pull/21387))
+- Allow PEP-604 unions in stubs and `TYPE_CHECKING` blocks prior to 3.10 ([#21379](https://github.com/astral-sh/ruff/pull/21379))
+- Ensure annotation/type expressions in stub files are always deferred ([#21401](https://github.com/astral-sh/ruff/pull/21401))
+- Further improve details around which expressions should be deferred in stub files ([#21456](https://github.com/astral-sh/ruff/pull/21456))
+- Sync vendored typeshed stubs ([#21466](https://github.com/astral-sh/ruff/pull/21466))
+
+### LSP server
+
+- Support for notebooks in VS Code ([#21175](https://github.com/astral-sh/ruff/pull/21175))
+- Fix goto for `float` and `complex` in type annotation positions ([#21388](https://github.com/astral-sh/ruff/pull/21388))
+- Add more keywords to scope-based completions ([#21383](https://github.com/astral-sh/ruff/pull/21383))
+- Add synthetic members to completions on dataclasses ([#21446](https://github.com/astral-sh/ruff/pull/21446))
+- Provide `import` completion when in `from <name> <name>` statement ([#21291](https://github.com/astral-sh/ruff/pull/21291))
+- Suppress completions when introducing names with `as` ([#21460](https://github.com/astral-sh/ruff/pull/21460))
+- Suppress invalid suggestions in import statements ([#21484](https://github.com/astral-sh/ruff/pull/21484))
+- Improve semantic token classification for names ([#21399](https://github.com/astral-sh/ruff/pull/21399))
+- Set `definition` modifier for parameter declarations when computing semantic tokens ([#21420](https://github.com/astral-sh/ruff/pull/21420))
+- Inlay hint call argument location ([#20349](https://github.com/astral-sh/ruff/pull/20349))
```

Comment:
Without looking at the PR, I'm not clear from this description what this actually does